### PR TITLE
refactor(timer): refactor timer

### PR DESCRIPTION
**Pull request by kherP**
- improve accessibility 
- introduce new props timerPrefix and timerSuffix
  - to allow user to pass in prefix/ suffix to be announced along with timer by screen reader

### CODE REVIEW

File: package-lock.json [hunk 1]
```diff
@@ -18439,7 +18439,182 @@
                 "treeverse",
                 "validate-npm-package-name",
                 "which",
-                "write-file-atomic"
+                "write-file-atomic",
```
Comment (eweseong) [inline package-lock.json:18442]: is the changes a patch from audit?

File: lib/src/Timer/Timer.tsx [hunk 2]
```diff
@@ -6,80 +6,67 @@ export type TimerProps = JSX.IntrinsicElements["div"] & {
...
-        this.clearInterval();
-    }
+    const convertToTwoDigits = (rawNumber: number): string => `${rawNumber < 10 ? "0" : ""}${rawNumber}`;
 
-    componentDidUpdate(prevProps: TimerProps): void {
-        if (prevProps.duration !== this.props.duration) {
-            if (this.props.duration !== null && this.props.duration !== undefined) {
-                this.startInterval(this.props.duration);
-            }
+    const convertMStoTime = (milliseconds: number): string => {
```
Comment (eweseong) [inline lib/src/Timer/Timer.tsx:45]: maybe we can rename this to `formatTime`? generally in date-time related libraries, they reserve `convertXX` for conversion between time units and `formatXyz` for display related transformation

Comment (eweseong) [reply lib/src/Timer/Timer.tsx:45]: or maybe even `formatMilitaryTime`

File: lib/src/Timer/Timer.tsx [hunk 3]
```diff
@@ -6,80 +6,67 @@ export type TimerProps = JSX.IntrinsicElements["div"] & {
...
+            clearInterval(innerInterval);
+            setInnerInterval(null);
         }
-    }
+    };
 
-    componentWillUnmount() {
-        this.clearInterval();
-    }
+    const convertToTwoDigits = (rawNumber: number): string => `${rawNumber < 10 ? "0" : ""}${rawNumber}`;
```
Comment (eweseong) [inline lib/src/Timer/Timer.tsx:43]: maybe we can use `.padStart` for this

```
const formatTwoDigits = (time: number): string => `${time}`.padStart(2, "0");
```

File: lib/src/Timer/Timer.tsx (hunk 2 above)
Comment (eweseong) [reply lib/src/Timer/Timer.tsx:45]: this algorithm seemed to work as well
```
return new Date(milliseconds).toISOString().slice(11, 19);
```

File: lib/src/Timer/Timer.tsx [hunk 4]
```diff
@@ -6,80 +6,67 @@ export type TimerProps = JSX.IntrinsicElements["div"] & {
...
-            if (this.props.duration !== null && this.props.duration !== undefined) {
-                this.startInterval(this.props.duration);
-            }
+    const convertMStoTime = (milliseconds: number): string => {
+        const rawSeconds: number = milliseconds / 1000;
+        const rawMinutes: number = rawSeconds / 60;
+        const displaySeconds: number = rawSeconds % 60;
+        const displayMinutes: number = Math.floor(rawMinutes % 60); // get remainder minutes
+        const displayHours: number = Math.floor(rawMinutes / 60); // get converted hours
+        return (displayHours > 0 ? displayHours + ":" : "") + convertToTwoDigits(displayMinutes) + ":" + convertToTwoDigits(displaySeconds);
```
Comment (eweseong) [inline lib/src/Timer/Timer.tsx:51]: the following code seemed to work as well
```
return new Date(milliseconds).toISOString().slice(11, 19);
```

Comment (kherP) [reply lib/src/Timer/Timer.tsx:51]: it will not work for duration more than 1 day

File: lib/src/Timer/Timer.tsx [hunk 5]
```diff
@@ -6,80 +6,67 @@ export type TimerProps = JSX.IntrinsicElements["div"] & {
...
-        };
-    }
-
-    startInterval(timeout: number): void {
-        this.setState({ timer: this.convertMStoTime(timeout) }, () => {
-            this.clearInterval();
-            this.innerInterval = setInterval(() => {
+    const startInterval = (timeout: number): void => {
+        setTimer(convertMStoTime(timeout));
+        clearInterval();
```
Comment (eweseong) [inline lib/src/Timer/Timer.tsx:21]: clear interval without interval id seem redundant, is this needed?

File: lib/src/Timer/Timer.tsx [hunk 6]
```diff
@@ -6,80 +6,67 @@ export type TimerProps = JSX.IntrinsicElements["div"] & {
...
-                    this.setState({ timer: this.convertMStoTime(timeout) }, () => {
-                        if (timeout === 0) {
-                            this.props.callback();
-                            this.clearInterval();
-                        }
-                    });
+                    setTimer(convertMStoTime(timeout));
+                    if (timeout === 0) {
+                        callback?.();
+                        clearInterval();
```
Comment (eweseong) [inline lib/src/Timer/Timer.tsx:29]: clear interval without interval id seem redundant, is this needed?

File: lib/src/Timer/Timer.tsx [hunk 7]
```diff
@@ -6,80 +6,67 @@ export type TimerProps = JSX.IntrinsicElements["div"] & {
...
+        const displaySeconds: number = rawSeconds % 60;
+        const displayMinutes: number = Math.floor(rawMinutes % 60); // get remainder minutes
+        const displayHours: number = Math.floor(rawMinutes / 60); // get converted hours
+        return (displayHours > 0 ? displayHours + ":" : "") + convertToTwoDigits(displayMinutes) + ":" + convertToTwoDigits(displaySeconds);
+    };
+
+    React.useEffect(() => {
+        if (duration >= 1000) {
+            // only start timer if there's at least 1 second set
+            clearInnerInterval();
```
Comment (eweseong) [inline lib/src/Timer/Timer.tsx:57]: is this step needed? am assuming on re-render/unmount the clean up code would be called?

File: package-lock.json (hunk 1 above)
Comment (kherP) [reply package-lock.json]: revert changes

File: lib/src/Timer/Timer.tsx (hunk 2 above)
Comment (kherP) [reply lib/src/Timer/Timer.tsx]: rename function

File: lib/src/Timer/Timer.tsx [hunk 8]
```diff
@@ -6,80 +6,67 @@ export type TimerProps = JSX.IntrinsicElements["div"] & {
...
-    componentDidMount() {
-        if (this.props.duration !== null && this.props.duration !== undefined) {
-            this.startInterval(this.props.duration);
+    const clearInnerInterval = (): void => {
+        if (innerInterval) {
+            clearInterval(innerInterval);
+            setInnerInterval(null);
         }
-    }
+    };
```
Comment (eweseong) [inline lib/src/Timer/Timer.tsx:41]: tried to debug the timer and notice the interval is still running after the duration has ended, could not pin point the root cause of it; but was able to clear the intervals after merging all the helper functions into the side effects
```
    React.useEffect(() => {
        let intervalId;

        if (duration >= 1000) {
            // only start timer if there's at least 1 second set
            setTimer(convertMStoTime(duration));
            intervalId = setInterval(() => {
                if (duration > 0) {
                    duration -= 1000;
                    setTimer(convertMStoTime(duration));
                } else {
                    callback?.();
                    clearInterval(intervalId);
                }
            }, 1000);
            setInnerInterval(intervalId);
        }

        return () => clearInterval(intervalId);
    }, [duration]);
```

Comment (kherP) [reply lib/src/Timer/Timer.tsx]: yes :(  i cleaned up the code

File: lib/src/Timer/Timer.tsx (hunk 5 above)
Comment (kherP) [reply lib/src/Timer/Timer.tsx]: should be clearInnerInterval

**Review by eweseong — APPROVED** [review]

**Review by eweseong — APPROVED** [review]

**Pull request merged**